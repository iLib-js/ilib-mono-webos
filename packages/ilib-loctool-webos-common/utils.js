/*
 * utils.js - Common utility functions for webos plugins are commonly used.
 *
 * Copyright (c) 2025 JEDLSoft
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var fs = require("fs");

/**
* Check whether a file exists at the given path.
*
* @param {String} filepath a path to file
*
* @returns {boolean} true if the path is valid; otherwise false.
*/
module.exports.isValidPath = function(filepath) {
    return filepath ? fs.existsSync(filepath) : false;
}

/**
* Read the content of a file from the specified path and parses it as JSON.
*
* @param {String} filepath a path to file

* @returns {Object} parsed JSON object if the path is valid; otherwise undefined.
*/
module.exports.loadData = function(filepath) {
    try {
        var readData = fs.readFileSync(filepath, 'utf-8');
        return JSON.parse(readData);
    } catch (error) {
        console.log(`Error reading or parsing file: ${error.message}`);
        return undefined;
    }
}

/**
* Check if the given key is exists in the file.
*
* @param {String} filepath a path to file
* @param {String} key The key to check for existence

* @returns {boolean} true if the is exists; otherwise false.
*/
module.exports.isExistKey = function(filepath, key) {
    if (!filepath || !key) return false;

    var jsonData = this.isValidPath(filepath) ? this.loadData(filepath) : {};
    return (jsonData && jsonData.hasOwnProperty(key));
}

/**
* Add a resource to this file
*
* @param {resFileType} resFileType a filetype of the resource to be added
* @param {Resource} translated a translation resource to add to this file
* @param {Resource} res a source resource for the translation resource
* @param {String} locale a resource of the locale
* @param {String} resPath localization resource path. It is optional
*
* @returns {boolean} true if the resource is added succesfully.
*/
module.exports.addResource = function (resFileType, translated, res, locale, resPath) {
    if (!resFileType || !translated || !res || !locale) return false;

    var file;
    // if reskeys don't match, we matched on cleaned string.
    // so we need to overwrite reskey of the translated resource to match
    if (translated.reskey !== res.reskey) {
        translated.reskey = res.reskey;
    }

    var resource = translated.clone();
    resource.project = res.getProject();
    resource.datatype = res.getDataType();
    resource.setTargetLocale(locale);
    resource.pathName = res.getPath();
    resource.context = res.getContext() || res.getPath().replace(/^.*[\\\/]/, '').replace(/\.(qml|js)/, "");
    file = resFileType.getResourceFile(locale, resPath);
    file.addResource(resource);

    return true;
}

/**
* Add a resource to this file
*
* @param {TranslationSet} newresSet translationset for the new resources
* @param {Resource} res a resource to add to this new translation set
* @param {String} locale the target locale of the resource
*
* @returns {boolean} true if the resource is successfully added.
*/
module.exports.addNewResource = function (newresSet, res, locale) {
    if (!newresSet || !res || !locale) return false;

    var note = "No translation for " + res.reskey + " to " + locale;
    var newres = res.clone();
    newres.setTargetLocale(locale);
    newres.setTarget(res.getSource());
    newres.setState("new");
    newres.setComment(note);
    newresSet.add(newres);

    return true;
};

/**
* @param {Object} settings settings object that configures how the tool will operate
*
* @returns {String} Information on which device is currently being targeted for localization
*/
module.exports.getDeviceType = function (settings) {
    if (!settings) return;
    return (settings && settings.metadata) ? settings.metadata["device-type"]: undefined;
}

/**
* @param {Resource} translated a translation resource to add to this file
* @param {String} deviceType It is optional. Information on which device is currently being targeted for localization
*
* @returns {String} The target string corresponding to the metadata.
*/
module.exports.getTarget = function (translated, deviceType) {
    var defaultTarget = translated.target;

    if (!deviceType || !translated.metadata) return defaultTarget;

    var dataArr = (translated.metadata && translated.metadata["mda:metaGroup"]) ?
                  translated.metadata["mda:metaGroup"]["mda:meta"]: undefined;

    if (!dataArr) return defaultTarget;
    else if (!Array.isArray(dataArr)) dataArr = [dataArr];

    var matchItem = dataArr.find(function(item) {
        return item['_attributes']['type'] === deviceType;
    });
    return matchItem ? matchItem['_text']: defaultTarget;
};