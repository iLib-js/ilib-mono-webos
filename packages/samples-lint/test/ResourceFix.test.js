/*
 * ResourceFix.test.js - test to check if the resource has been fixed
 *
 * Copyright (c) 2025 JEDLSoft
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import path from 'path';
import { loadFileData, isValidPath } from "./utils.js";

const findTarget = (units, srcString) => {
    for (const unit of units) {
        if (unit?.segment?.source?._text === srcString) {
            return unit.segment.target?._text ?? '';
        }
    }
    return '';
};

describe('test if the resource is modifed properly', () => {
    let filePath, result;
    test("test resource-edge-whitespace (am-ET)", function() {
        expect.assertions(3);
        const source = "You can use your Magic Remote to determine the room’s acoustics and optimize your TV’s sound quality based on the results."
        const target = "MAGIC REMOTE በመጠቀም የክፍሉን የድምጽ ማስተጋባት አና በውጤቶቹ መሰረት የቲቪዎን የድምጽ ጥራት ማሰተካከል እና መቆጣጠር ይችላሉ፡፡";

        filePath = path.join('sample-rs-edge-whitespace', 'am-ET.xliff.modified'); // autofix
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        expect(xliffData).toBeTruthy();

        result = findTarget(xliffData.group.unit, source);
        expect(result).toBe(target);

        filePath = path.join('sample-rs-edge-whitespace', 'am-ET.xliff'); //original
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        result = findTarget(xliffData.group.unit, source);
        expect(result).not.toBe(target);
    });

    test("test am-ET.xliff.modified file", function() {
        expect.assertions(1);

        filePath = path.join('sample-rs-edge-whitespace', 'am-ET.xliff.modified'); // autofix
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        expect(xliffData).toMatchSnapshot();
    });

    test("test resource-edge-whitespace (zh-Hans-CN) 1", function() {
        expect.assertions(3);
        const source = "Select your remote's cursor speed and appearance."
        const target = "选择您遥控器的光标速度和外观。";

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff.modified'); // autofix
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        expect(xliffData).toBeTruthy();

        result = findTarget(xliffData.group.unit, source);
        expect(result).toBe(target);

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff'); //original
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        result = findTarget(xliffData.group.unit, source);
        expect(result).not.toBe(target);
    });

    test("test resource-edge-whitespace (zh-Hans-CN) 2", function() {
        expect.assertions(3);
        const source = "press OK button to go to next menu"
        const target = "按“确定”按钮以前往下一个菜单";

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff.modified'); // autofix
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        expect(xliffData).toBeTruthy();

        result = findTarget(xliffData.group.unit, source);
        expect(result).toBe(target);

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff'); //original
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        result = findTarget(xliffData.group.unit, source);
        expect(result).not.toBe(target);
    });

    test("test resource-edge-whitespace (zh-Hans-CN) 3", function() {
        expect.assertions(3);
        const source = "If the image isn't clear, select another pattern until the picture improves."
        const target = "如果图像不清晰，请选择另一模式，直至图片更加清晰。";

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff.modified'); // autofix
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        expect(xliffData).toBeTruthy();

        result = findTarget(xliffData.group.unit, source);
        expect(result).toBe(target);

        filePath = path.join('sample-rs-edge-whitespace', 'zh-Hans-CN.xliff'); //original
        xliffData = isValidPath(filePath) ? loadFileData(filePath) : {};
        result = findTarget(xliffData.group.unit, source);
        expect(result).not.toBe(target);
    });
});