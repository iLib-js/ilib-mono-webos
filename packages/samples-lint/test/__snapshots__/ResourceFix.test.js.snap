// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`test if the resource is modifed properly test am-ET.xliff.modified file 1`] = `
{
  "_attributes": {
    "id": "sample-rs-edge-whitespace_f1",
    "original": "sample-rs-edge-whitespace",
  },
  "group": {
    "_attributes": {
      "id": "sample-rs-edge-whitespace_g1",
      "name": "javascript",
    },
    "unit": [
      {
        "_attributes": {
          "id": "sample-rs-edge-whitespace_g1_1",
        },
        "segment": {
          "source": {
            "_text": "{%1} hour(s)",
          },
          "target": {
            "_text": "{%1} ሰዓት(ታት)",
          },
        },
      },
      {
        "_attributes": {
          "id": "sample-rs-edge-whitespace_g1_2",
        },
        "segment": {
          "source": {
            "_text": "{%1} min(s)",
          },
          "target": {
            "_text": "{%1} ደቂቃ(ዎች)",
          },
        },
      },
      {
        "_attributes": {
          "id": "sample-rs-edge-whitespace_g1_3",
        },
        "segment": {
          "source": {
            "_text": "Loading...",
          },
          "target": {
            "_text": "በመጫን ላይ...",
          },
        },
      },
      {
        "_attributes": {
          "id": "sample-rs-edge-whitespace_g1_4",
        },
        "segment": {
          "source": {
            "_text": "You can use your Magic Remote to determine the room’s acoustics and optimize your TV’s sound quality based on the results.",
          },
          "target": {
            "_text": "MAGIC REMOTE በመጠቀም የክፍሉን የድምጽ ማስተጋባት አና በውጤቶቹ መሰረት የቲቪዎን የድምጽ ጥራት ማሰተካከል እና መቆጣጠር ይችላሉ፡፡",
          },
        },
      },
    ],
  },
}
`;
